### ISSUE TITLE
Download Failures

### ISSUE THREAD

**Issue by sim099** [gilesknap/gphotos-sync#236]
I emailed Giles about an issue, as I wasn't certain it was a bug and wanted to ask about it before raising it here.

He asked me to raise a ticket, so I've pasted the relevant content from my email below.

The scenario:
 * I have two google accounts and have synchronized the same video files to them both via google photos on my mobile
 * All photos and videos have downloaded successfully for the first account
 * A number of video files won't download for the second account due to internal server errors
 * These are all videos that have been uploaded in the last few days

I did temporarily see this behaviour when attempting to synchronize the first account too, but I tried again a few times and it just started working. I've tried a lot more than that with the second account and still haven't had any success.

I did notice the "skipped files" message on screen and taken a look at the gphotos.bad_ids.yaml file. I noticed that the URLs listed in this file are very different to those output on screen and am not sure if this is expected or not.

This is the first URL listed in the bad_ids file:

https://photos.google.com/lr/photo/ABGoW6Y-LS__WbMiSEcxjySMGeLd16kFvUKDpK-gFv3YmJJQYdm77D2BMVKFgeUXKsckdmrZNxBUhHBOwDSVgpDujamNz0QM1g

I pasted this into my browser and video loads fine.

I copied the video filename via the information tab and searched the log file for related data. This is the related block:

05-19 17:58:06 gphotos.GooglePhotosDownload ERROR    FAILURE 17 downloading photos\2020\05\brltw8xl0glkujbe.mp4 - 500 Server Error: Internal Server Error for url: https://video-downloads.googleusercontent.com/AGQNM9Lbtww8W2pb0tUKOpwUk2YNRVj0DowHgRfZfpXBhuqJyuiX34ITcrHkao8nsq0zDL0jO4jnSITo01dMNIb3Ed_aPe9aJUg4J6grrK1QhB3rOU3sSFX6J5dMKaO6MVU5KT4jAIRQrU9s117rQp7POAkoGKlWrHY95e7gQxpfgvBb4pVwr5N3X4_1vjrIeGH-KEbMJRyfyFaxUkKZb8D2TGWhlihQsED1WzUUq7ZJm_eoFpNQHkRSGsdo0LDO9UZmJFdOFpxHpb0CwwrgK4JMj_wT7jGttZhzUU5fGYWoWA4N1Q-Fh2TiLKWhxAebYY5kd_CLoiiY0KzmpthnQH0ajjYqNcENUbHdsU1ustV_OUtCs4ufCL4w9OYgR2R-HLk_gMQY7_Jht2YzBc2uoG0FZAbXtB9_jtPhX029XjZZr85uT02mKmUIUhhbhOStlb440OmWzWOJ7bUmQhI-zgehI1PCuDrsfYDst99LtavUo5mWuJNhAPOYNOEba9X2TxyeLESz3zO2OGa5ehBOj_jynzEPNmptAYh8cCPLvVkvUxhezxxe1cQIoS9PtG-hoSlHnoQobQT-BQn82M0OflAyjt8OlrUU_bwSz2Yz1XAl0gHiFJnn6RnVelHUqa_d_lnALl5A_CcmwHXcwXIfaDzYIgog1sV-d07Kua3s8NHHVWxdeyM7zU9_HFc8CYfZvn7QU24l_36OoUgBSuLVBxCSjtXwmWhQT3VVI10ZHsNEbzKjZdeHBE1g7mKuvjKCRBdMeHLtCPvN2bBEJu4-v9oqTvgPqpo7Duv2YIXUA68sHJn1f0tLsVIAmIPHdEvMjNPwjzvOj-92mrv4jaZ414864eDNEbMRVgo37z78_NqhVlo9M6iXWeXwKSm96TmVR_kUbX4hWX3BF0FKZhu1c_9UMNqxlMoVww
05-19 17:58:06 gphotos.BadIds DEBUG    BAD ID ABGoW6Y-LS__WbMiSEcxjySMGeLd16kFvUKDpK-gFv3YmJJQYdm77D2BMVKFgeUXKsckdmrZNxBUhHBOwDSVgpDujamNz0QM1g for photos\2020\05\brltw8xl0glkujbe.mp4
Traceback (most recent call last):
  File "C:\applications\WinPython64\WPy64-3810\python-3.8.1.amd64\lib\concurrent\futures\thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "c:\applications\winpython64\wpy64-3810\settings\.virtualenvs\notebooks-ubarzqfp\lib\site-packages\gphotos\GooglePhotosDownload.py", line 256, in do_download_file
    response.raise_for_status()
  File "c:\applications\winpython64\wpy64-3810\settings\.virtualenvs\notebooks-ubarzqfp\lib\site-packages\requests\models.py", line 941, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 500 Server Error: Internal Server Error for url: https://video-downloads.googleusercontent.com/AGQNM9Lbtww8W2pb0tUKOpwUk2YNRVj0DowHgRfZfpXBhuqJyuiX34ITcrHkao8nsq0zDL0jO4jnSITo01dMNIb3Ed_aPe9aJUg4J6grrK1QhB3rOU3sSFX6J5dMKaO6MVU5KT4jAIRQrU9s117rQp7POAkoGKlWrHY95e7gQxpfgvBb4pVwr5N3X4_1vjrIeGH-KEbMJRyfyFaxUkKZb8D2TGWhlihQsED1WzUUq7ZJm_eoFpNQHkRSGsdo0LDO9UZmJFdOFpxHpb0CwwrgK4JMj_wT7jGttZhzUU5fGYWoWA4N1Q-Fh2TiLKWhxAebYY5kd_CLoiiY0KzmpthnQH0ajjYqNcENUbHdsU1ustV_OUtCs4ufCL4w9OYgR2R-HLk_gMQY7_Jht2YzBc2uoG0FZAbXtB9_jtPhX029XjZZr85uT02mKmUIUhhbhOStlb440OmWzWOJ7bUmQhI-zgehI1PCuDrsfYDst99LtavUo5mWuJNhAPOYNOEba9X2TxyeLESz3zO2OGa5ehBOj_jynzEPNmptAYh8cCPLvVkvUxhezxxe1cQIoS9PtG-hoSlHnoQobQT-BQn82M0OflAyjt8OlrUU_bwSz2Yz1XAl0gHiFJnn6RnVelHUqa_d_lnALl5A_CcmwHXcwXIfaDzYIgog1sV-d07Kua3s8NHHVWxdeyM7zU9_HFc8CYfZvn7QU24l_36OoUgBSuLVBxCSjtXwmWhQT3VVI10ZHsNEbzKjZdeHBE1g7mKuvjKCRBdMeHLtCPvN2bBEJu4-v9oqTvgPqpo7Duv2YIXUA68sHJn1f0tLsVIAmIPHdEvMjNPwjzvOj-92mrv4jaZ414864eDNEbMRVgo37z78_NqhVlo9M6iXWeXwKSm96TmVR_kUbX4hWX3BF0FKZhu1c_9UMNqxlMoVww

If I attempt to load this URL via my browser then I get an error 400, but I suspect that I'm not supposed to be able to access that via a browser.

Is there anything that I'd be able to do in order to try and work out what's happening here?

**Comment by sim099** [comment]
This was his response:

> I don't know why Google is presenting with 500 errors but this does seem to have happened more often recently. We used to have a bug in their API that was fixed some time ago and something new has turned up or there has been a regression. I would guess that having two accounts is not related but anything is possible.
> 
> Because of this I restored the 'Bad Ids' functionality. Now if the issue was transient the bad ids file is detrimental since it stops the file from being retried.
> 
> So the first thing to try is deleting the bad ids file and then running again with --rescan.
> 
> Lets see where you get with that.
> 
> If you can make a ticket on gitub for your response then it helps others who are seeing the same issue.

I hadn't mentioned in my email that I've been using the following command line each time I execute the script:
`call %winpython3_home%\scripts\winpython -m pipenv run gphotos-sync --secret "{path to json file} --flush-index --do-delete --skip-albums "{destination path}"`

I initially did this because I was synchronising the two accounts and there were a lot of diferences, so I was making some changes (including deletions) and then re-syncing and doing a directory comparison locally. I suppose I could update my batch file to stop flushing and deleting unless I know it's needed now.

I have carried out the requested actions and executed with --rescah after removing gphotos.bad_ids.yaml. Unfortunately, it failed on the same 25 videos.

**Comment by gilesknap** [comment]
Let's get a log of exactly what is happening in the API.

Please  can you re-run with ```--log-level trace```, then send me the files ending in .trace and .log from the destination path. The files might be big so you can send them directly to me via file.io or dropbox.

Before the run, please delete the bad ids file and run without ```--flush-index```, but with ```--rescan```, since this will keep the filesize down a little.

**Comment by sim099** [comment]
I've just done so and ended up with everything downloading successfully, so I've got no idea what was preventing them from doing so before.

I did take a look at the trace information though. There's a lot of data being processed!

Just out of interest, the productUrl is the one that I can use to access an item. Is that the one that's used withing the script when downloading a file or is baseUrl used?

**Comment by gilesknap** [comment]
So this appears to be transient issues on the google server. In the past I have found that google may spend some time processing a video that was recently uploaded and that it would not be possible to download for some time. This was before I had gphotos-sync and just through the web gui. Maybe it was something like that? Were the videos quite large?

The productUrl takes you to the web gui page for that media item. It is baseUrl that is used for downloading the data itself. 

Closing as not an issue in gphotos-sync.

**Comment by sim099** [comment]
No, they were max 20mb in size.

I know what you mean about the processing and I have it set to allow unlimited uploads in HD quality, so I've uploaded large videos (1-2gb) and then had a smaller version (few humdred mb) download via gphotos-sync. They must process all video files to some degree, as there's a "video is processing" delay for much smaller files (few mb, standard definition) too.

In this case though, I could view the video via the URL I tested, so I know it had been fully processed at that point.

It doesn't matter though, as you say it can't be an issue in gphotos-sync seeing as the files eventually downloaded without anything changing on the client side.

So the productUrl provides a data stream and would require authentication headers. That explains wht it woudn't load in a browser. :-)

Thanks!
